src/PowerShellEditorServices.Hosting/NamedPipeUtils.cs

```diff
@@ -0,0 +1,154 @@
+            var pipeSecurity = new PipeSecurity();
+
+            WindowsIdentity identity = WindowsIdentity.GetCurrent();
+            WindowsPrincipal principal = new WindowsPrincipal(identity);
+
+            if (principal.IsInRole(WindowsBuiltInRole.Administrator))
+            {
+                // Allow the Administrators group full access to the pipe.
+                pipeSecurity.AddAccessRule(
+                    new PipeAccessRule(
+                        new SecurityIdentifier(WellKnownSidType.BuiltinAdministratorsSid, domainSid: null).Translate(typeof(NTAccount)),
+                        PipeAccessRights.FullControl, AccessControlType.Allow));
+            }
+            else
+            {
+                // Allow the current user read/write access to the pipe.
+                pipeSecurity.AddAccessRule(new PipeAccessRule(
+                    WindowsIdentity.GetCurrent().User,
+                    PipeAccessRights.ReadWrite, AccessControlType.Allow));
+            }
+
+            return new NamedPipeServerStream(
+                pipeName,
+                pipeDirection,
+                maxNumberOfServerInstances: 1,
+                PipeTransmissionMode.Byte,
+                PipeOptions.Asynchronous,
+                inBufferSize: PipeBufferSize,
+                outBufferSize: PipeBufferSize,
+                pipeSecurity);
+#endif
+        }
+
+        /// <summary>
+        /// Generate a named pipe name known to not already be in use.
+        /// </summary>
+        /// <param name="prefixes">Prefix variants of the pipename to test, if any.</param>
+        /// <returns>A named pipe name or name suffix that is safe to you.</returns>
+        public static string GenerateValidNamedPipeName(IReadOnlyCollection<string> prefixes = null)
+        {
+            int tries = 0;
+            do
+            {
+                string pipeName = $"PSES_{Path.GetRandomFileName()}";
+
+                // In the simple prefix-less case, just test the pipe name
+                if (prefixes == null)
+                {
+                    if (!IsPipeNameValid(pipeName))
+                    {
+                        continue;
+                    }
+
+                    return pipeName;
+                }
+
+                // If we have prefixes, test that all prefix/pipename combinations are valid
+                bool allPipeNamesValid = true;
+                foreach (string prefix in prefixes)
+                {
+                    string prefixedPipeName = $"{prefix}_{pipeName}";
+                    if (!IsPipeNameValid(prefixedPipeName))
+                    {
+                        allPipeNamesValid = false;
+                        break;
+                    }
+                }
+
+                if (allPipeNamesValid)
+                {
+                    return pipeName;
+                }
+
+            } while (tries < 10);
+
+            throw new Exception("Unable to create named pipe; no available names");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/app/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=4579423)